packages/arb-bridge-peripherals/contracts/tokenbridge/ethereum/gateway/L1ArbitrumGateway.sol

```diff
@@ -0,0 +1,180 @@
+    function isCounterpartGateway() internal view virtual override returns (bool) {
+        IOutbox outbox = IOutbox(IInbox(inbox).bridge().activeOutbox());
+        return counterpartGateway == outbox.l2ToL1Sender();
+    }
+
+    function _initialize(
+        address _l2Counterpart,
+        address _router,
+        address _inbox
+    ) internal virtual {
+        // L1 gateway must have a router
+        require(_router != address(0), "BAD_ROUTER");
+        TokenGateway._initialize(_l2Counterpart, _router);
+        require(_inbox != address(0), "BAD_INBOX");
+        router = _router;
+        inbox = _inbox;
+    }
+
+    /**
+     * @notice Finalizes a withdrawal via Outbox message; callable only by L2Gateway.outboundTransfer
+     * @param _token L1 address of token being withdrawn from
+     * @param _from initiator of withdrawal
+     * @param _to address the L2 withdrawal call set as the destination.
+     * @param _amount Token amount being withdrawn
+     * @param _data encoded exitNum (Sequentially increasing exit counter determined by the L2Gateway) and additinal hook data
+     */
+    function finalizeInboundTransfer(
+        address _token,
+        address _from,
+        address _to,
+        uint256 _amount,
+        bytes calldata _data
+    ) external virtual override onlyCounterpartGateway returns (bytes memory) {
+        (uint256 exitNum, bytes memory extraData) = abi.decode(_data, (uint256, bytes));
+
+        // TODO: add withdraw and call
+        // TODO: add transferExit
+        IERC20(_token).safeTransfer(_to, _amount);
+        emit InboundTransferFinalized(_token, _from, _to, exitNum, _amount, _data);
+
+        return bytes("");
+    }
+
+    function createOutboundTx(
+        address _user,
+        uint256 _maxSubmissionCost,
+        uint256 _maxGas,
+        uint256 _gasPriceBid,
+        bytes memory _data
+    ) internal virtual returns (uint256) {
+        // msg.value is sent, but 0 is set to the L2 call value
+        // the eth sent is used to pay for the tx's gas
+        uint256 seqNum =
+            IInbox(inbox).createRetryableTicket{ value: msg.value }(
+                counterpartGateway,
+                0,
+                _maxSubmissionCost,
+                _user,
+                _user,
+                _maxGas,
+                _gasPriceBid,
+                _data
+            );
+        return seqNum;
+    }
+
+    /**
+     * @notice Deposit ERC20 token from Ethereum into Arbitrum. If L2 side hasn't been deployed yet, includes name/symbol/decimals data for initial L2 deploy. Initiate by GatewayRouter.
+     * @param _token L1 address of ERC20
+     * @param _to account to be credited with the tokens in the L2 (can be the user's L2 account or a contract)
+     * @param _amount Token Amount
+     * @param _maxGas Max gas deducted from user's L2 balance to cover L2 execution
+     * @param _gasPriceBid Gas price for L2 execution
+     * @param _data encoded data from router and user
```

Comment:
specing out the path (either here or elsewhere) of what data gets encoding and where/when could be useful, as it's a bit confusing to follow